Add ability to generate product-dependencies from published recommendations

jared2501 opened the pull request:
Adds the ability for the CreateManifestTask to infer the min/max/recommended versions for product dependencies by reading "pdep" files that API publishers can include alongside their published jar files. The "prdep" file includes a list of product dependencies that a consumer of the jar should take.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/sls-packaging/242)
<!-- Reviewable:end -->

Commit:
Add detectConstraints flag to ProductDependency

Commit:
Lazily compute product dependencies in CreateManifestTask

Commit:
Hook up product recommendations in CreateManifestTask

jared2501 commented:
@uschi2000 @gracew @jmcampanini @markelliot - this is part 1 of 2 of what we discussed. Part 2 will add a task to generate pdep files.

jared2501 commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +1,6 @@
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-3.4.1-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.1-bin.zip
```
needed to upgradle so that the GradleDependencyGenerator worked

jmcampanini replied:
nit: use `-all` version of the distribution for gradle plugins, so the source code is present locally

markelliot reviewed (commented):
So I think we don't need pdep files and instead can put this info in jar manifests -- when the jars show up on the classpath, one should only need to include the jars rather than also explicitly list the group/name parts.

jared2501 commented:
@markelliot - Hmm yeah, I discussed this option with @uschi2000 and decided against it since pdep files can work with non-java languages as well. No strong prefs either way though.

Commit:
Reduce gradle memory usage

jmcampanini commented:
having the information in the POM doesn't limit it to jvm-languages - it just limits it to a maven layout + repository, which is almost a requirements if we're using `pdeps` that are being published (technically you could publish `pdeps` anywhere).

you can access the POM data from artifactory with just a link: 

```
https://artifactory.server/artifactory/my-repo-name/com/palantir/package/artifact//0.1.0/artifact-0.1.0.pom
```

you just end up with an XML file, which may be more tedious to parse in a non-jvm language (more so than JSON or just plaintext or property file)

i'd say the pros/cons for `pdeps` over `POM` would be:

* PRO: the ability to publish + host from a different server/repo, no longer requiring artifactory
* PRO: may be slightly easier parsing of file
* PRO: provides a non-invasive path for adding to already-published artifacts (additive vs replacing existing POM)
* CON: POM files are already the canonical source of dependency information. arguably this is a different kind of dependency, so this is more of a silly con

jared2501 commented:
Also, using POM scopes for this information is probably an anti-pattern since the behaviour if a dependency is in multiple scopes is super confusion (see https://discuss.gradle.org/t/gradle-1-9-dependency-resolution-issue-if-artifact-present-in-multiple-scopes-within-the-pom/2053/6).

jmcampanini commented on build.gradle:
```diff
@@ -12,6 +12,9 @@ buildscript {
     }
 }
+plugins {
+    id 'org.inferred.processors' version '1.1.3'
```
swap to buildscript method? (to follow existing convention)

jared2501 replied:
done

jmcampanini commented on circle.yml:
```diff
@@ -3,7 +3,7 @@ machine:
   environment:
     TERM: dumb
-    GRADLE_OPTS: '-Dorg.gradle.jvmargs="-Xmx2048m -XX:+HeapDumpOnOutOfMemoryError"'
+    GRADLE_OPTS: '-Dorg.gradle.jvmargs="-Xmx1024m -XX:+HeapDumpOnOutOfMemoryError"'
```
curious: why half the size for this PR?

jared2501 replied:
the circle builds were OOMing since free version only gives you 1GB .<

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GradleTestSpec.groovy:
```diff
@@ -16,34 +16,45 @@
                 .withProjectDir(projectDir)
-                .withArguments(tasks)
+                .withArguments(ImmutableList.<String> builder()
+                .addAll(Arrays.asList(tasks))
```
nit: indent is funky/deceiving

jared2501 replied:
intellij auto-formats as such, will change though

jared2501 replied:
done

uschi2000 commented:
I think we explicitly don't want transitive resolution?

Sent from Email+ secured by MobileIron

-------- Original Message --------
From: jared2501 <notifications@github.com<mailto:notifications@github.com>>
Date: Tue, Aug 15, 2017, 2:09 PM
To: palantir/sls-packaging <sls-packaging@noreply.github.com<mailto:sls-packaging@noreply.github.com>>
CC: Robert Fink <rfink@palantir.com<mailto:rfink@palantir.com>>,Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: Re: [palantir/sls-packaging] Add ability to generate product-dependencies from published recommendations (#242)

Also, using POM scopes for this information is probably an anti-pattern since the behaviour if a dependency is in multiple scopes is super confusion (see https://discuss.gradle.org/t/gradle-1-9-dependency-resolution-issue-if-artifact-present-in-multiple-scopes-within-the-pom/2053/6[discuss.gradle.org]<https://urldefense.proofpoint.com/v2/url?u=https-3A__discuss.gradle.org_t_gradle-2D1-2D9-2Ddependency-2Dresolution-2Dissue-2Dif-2Dartifact-2Dpresent-2Din-2Dmultiple-2Dscopes-2Dwithin-2Dthe-2Dpom_2053_6&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=UfcWCaR4ui50AFap-gezrx5XYtPtH-9JpazU7tbRW-4&m=tjuD5Gix14amP4ZkiqrKra_VfU9qTY7ZIATa91mEBGs&s=D8wQFVok-C8nvyny_EZrM-YEEWE1UA-2g9A1Vv7awUw&e=>).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_sls-2Dpackaging_pull_242-23issuecomment-2D322559867&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=UfcWCaR4ui50AFap-gezrx5XYtPtH-9JpazU7tbRW-4&m=tjuD5Gix14amP4ZkiqrKra_VfU9qTY7ZIATa91mEBGs&s=KPDwv-6agceCz9JNo2hXkcmfwX6k2hLhl-u0kIiZ3jg&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AGOdwddBVSwKwtn62Y8idoYzCVj3ksF5ks5sYezUgaJpZM4O3CM4&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=UfcWCaR4ui50AFap-gezrx5XYtPtH-9JpazU7tbRW-4&m=tjuD5Gix14amP4ZkiqrKra_VfU9qTY7ZIATa91mEBGs&s=h324mIvthRbZJzi1u3iLiRPA2m-AKRD-TWw6Dq5dFzU&e=>.

markelliot commented:
You do, if you pick up an API jar via a library dependency, you should pick up the product dependency. 

My vote was for the literal jars and to look at the classpath, rather than manage through POMs. While this won't work for all API deps of all languages, for Java services it should be fine, and this is primarily a packager for Java services. 

We should talk through the assets case separately, as we likely want to use package.json to be more npm-centric (but you kind of get the thing I'm going for).

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,159 @@
+            }
+
+            repositories {
+                maven {url "file:///{{mavenRepo}}"}
```
if you use `"""` instead of `'''` you can use string interpolation and switch this to `${mavenRepo.getAbsolutePath()}`

jared2501 replied:
done

uschi2000 commented:
It's likely a separate topic, but I suspect that shared libraries with extensive dependencies on service jars are an antipattern.

In today's reality (=100% of services are in Java), the JAR manifest thing could work. Not super future-proof,  but likely the solution that's gets you near-term wins most quickly.

Sent from Email+ secured by MobileIron

-------- Original Message --------
From: Mark Elliot <notifications@github.com<mailto:notifications@github.com>>
Date: Tue, Aug 15, 2017, 2:23 PM
To: palantir/sls-packaging <sls-packaging@noreply.github.com<mailto:sls-packaging@noreply.github.com>>
CC: Robert Fink <rfink@palantir.com<mailto:rfink@palantir.com>>,Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Subject: Re: [palantir/sls-packaging] Add ability to generate product-dependencies from published recommendations (#242)

You do, if you pick up an API jar via a library dependency, you should pick up the product dependency.

My vote was for the literal jars and to look at the classpath, rather than manage through POMs. While this won't work for all API deps of all languages, for Java services it should be fine, and this is primarily a packager for Java services.

We should talk through the assets case separately, as we likely want to use package.json to be more npm-centric (but you kind of get the thing I'm going for).

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_palantir_sls-2Dpackaging_pull_242-23issuecomment-2D322563328&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=UfcWCaR4ui50AFap-gezrx5XYtPtH-9JpazU7tbRW-4&m=OiXInPsbOWz6lZ4Z3vV7i-1D_qQ6PXJPO93ZPDA7qqs&s=VUUeDEsiby8i7u11QuYdiedz0g2H2Nm2SbUy9IWssSo&e=>, or mute the thread[github.com]<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AGOdwWCwhSJJ-2DsRdxG70a7Oy1Ltyl4iqks5sYfA5gaJpZM4O3CM4&d=DwMFaQ&c=izlc9mHr637UR4lpLEZLFFS3Vn2UXBrZ4tFb6oOnmz8&r=UfcWCaR4ui50AFap-gezrx5XYtPtH-9JpazU7tbRW-4&m=OiXInPsbOWz6lZ4Z3vV7i-1D_qQ6PXJPO93ZPDA7qqs&s=V-TedtONo4OsiDmX3UPGFRTNOFpYOHJUYdzCwzvp9j0&e=>.

jared2501 commented:
In the jar manifest world, what would you do about asset/asset bundles dependencies? I.e. spark module -> spark dist asset bundle dependencies?

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,159 @@
+                productType = "service"
+                manifestExtensions = [:]
+                productDependencies = []
+                productDependenciesConfig = configurations.runtime
```
could we not assume this will always be runtime? at least set it by default?

jared2501 replied:
runtime is created by the java plugin, asset bunldes don't have the java plugin applied

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,159 @@
+                productDependencies = []
+                productDependenciesConfig = configurations.runtime
+                ignoredProductIds = [
+                    new com.palantir.gradle.dist.ProductId("group:name"), 
```
can these be just strings and have the plugin do the conversion to `ProductId` ?

jared2501 replied:
we could, but it appears the convention of this code base is to have the task take the typed things and do the conversion in the extension - happy to change this though

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,159 @@
+                productType = "service"
+                manifestExtensions = [:]
+                productDependencies = [
+                    new com.palantir.gradle.dist.ProductDependency("group", "name"),
```
1. could we use strings here instead as well?
2. could this use a dependency configuration: `slsProduct` or something similar? (so it would use `runtime` + `slsProduct`, or `slsProduct` extends runtime)

jared2501 replied:
1. see comment above
2. think productDependencies should not move over to a configuration since you may want to explicitly list the ProductDependencies and override the suggested min/max/recommended versions

jmcampanini reviewed (commented):
not a full review, attempting to understand the tests first

jared2501 commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,159 @@
+                productType = "service"
+                manifestExtensions = [:]
+                productDependencies = [
+                    new com.palantir.gradle.dist.ProductDependency("group", "name"),
```
see comment above

Commit:
address comments

jared2501 commented:
@jmcampanini - addressed your comments.

@uschi2000 / @markelliot - happy to change to using the jar manifest, but would like an ACK that we're OK forgoing the ability for spark module asset bundles (which do use gradle + maven) to use this feature.

uschi2000 commented:
I have a slight preference for not using the JAR approach *if we don't need transitives*. Mark, if you think transitives are required, then I'm fine with the JAR thing too.

uschi2000 commented:
OK, misinterpreted how you resolve&fetch the pdep files.

uschi2000 commented:
Would be interested in understanding how this will interface with authoring conjure APIs though?

jared2501 commented:
I haven't done that part yet, this part only consumes the pdep files. I imagine you'd want that part to look like this:

```
// stemma-api/build.gradle
apply plugin "sls-packaging-recommended-product-dependencies"

recommendedProductDependencies {
  recommendedProductDependency {
    productGroup = "com.palantir.stemma"
    productName = "stemma"
    minVersion = "1.0.0"
    maxVersion = "1.X.X"
    recommendedVersion = "1.2.3"
  }
}
```

jared2501 commented:
Ping here - is this OK to move forward with?

jared2501 commented:
Talked with @markelliot offline and decided that the asset <-> asset bundle case can probably be handled with jar manifests, and the upside of the manifest approach is that changing what people publish is harder than shoving stuff in their jars.

Will change this PR to use the jar manifests instead of pdep files.

Commit:
Remove pdep files and use jar manifest instead

jared2501 commented:
@jmcampanini - made the change to jar manifests, ready for your review :)

jmcampanini commented on build.gradle:
```diff
@@ -5,13 +5,17 @@ buildscript {
     }
 }
+plugins {
+    id 'org.inferred.processors' version '1.1.3'
```
would switch to an apply statement

jared2501 replied:
woops forgot to remove this, done

jmcampanini commented on .gitignore:
```diff
@@ -32,3 +31,5 @@ fileeditor
 .idea/
 out/
+
+**/generated_src/
```
nit: add test as well, to help future-you

jared2501 replied:
what test would I add for this? not sure if it's worth future proofing git ignore file

jmcampanini commented on gradle-sls-packaging/build.gradle:
```diff
@@ -1,15 +1,18 @@
 
     testCompile gradleTestKit()
-    testCompile "com.netflix.nebula:nebula-test:4.0.0" // for better temp directory junit rule only
+    testCompile "com.netflix.nebula:nebula-test:6.1.2"
```
single quotes

jared2501 replied:
done

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/GradleTestSpec.groovy:
```diff
@@ -16,34 +16,45 @@
 import groovy.transform.CompileStatic
 import groovy.transform.TypeCheckingMode
 import nebula.test.multiproject.MultiProjectIntegrationHelper
+import org.gradle.internal.impldep.com.google.common.collect.ImmutableList
```
just-comment: in test where we control the classpath, less necessary to use a shaded version of a guava class

jared2501 replied:
woops, IJ auto import failure

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,156 @@
+
+class CreateManifestTaskTest extends GradleTestSpec {
+
+    def jsonMapper = new ObjectMapper()
```
nit: can use explicit type `ObjectMapper` here

jmcampanini replied:
can we ask the `CreateManifestTask` class to create an objectmapper to not duplicate the code?

jared2501 replied:
done

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,156 @@
+        then:
+        buildResult.task(':testCreateManifest').outcome == TaskOutcome.FAILED
+        buildResult.output.contains("The following products are recommended as dependencies but do not appear in the " +
+                "product dependencies or product dependencies ignored list: [group:name2, group:name]")
```
how does it know it needs to depend on `group:name2` and `group:name`?

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/tasks/CreateManifestTaskTest.groovy:
```diff
@@ -1,10 +1,156 @@
+                manifestExtensions = [:]
+                productDependencies = []
+                productDependenciesConfig = configurations.runtime
+                ignoredProductIds = [
```
why not use: `ignoredProductDependency(String productGroup, String productName)`?

jared2501 replied:
The runtime configuration brings in `a:a:1.0`, the test makes it so that `a:a:1.0` depends on `b:b:1.0` as well and both `a:a:1.0` and `b:b:1.0` have stuff jammed in the MANIFEST (I checked it into the test repo)

jmcampanini replied:
i meant why not call the method on the extension to add things to the `Set` rather than setting the `Set` explicitly

jared2501 replied:
ah - cause this test was testing just the task separately (i.e. this test creates it's own task "testCreateManifest" instead of using the conventional "createManifest")... I could change all the tests do use "createManifest" and then invoke the task and use "-x downStreamTask1 -x downStreamTask2, etc", but decided against cause im lazy >.<

jmcampanini replied:
cool.

assuming there only 1 manifest per project, i'd say that providing an extension to override is simpler/nicer than customizing a task. but that's ergonomic related, not functionality related, so can be a separate PR if desired

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductDependency.groovy:
```diff
@@ -1,11 +1,24 @@
-class ProductDependency {
+@ToString
+@EqualsAndHashCode
+class ProductDependency implements Serializable {
```
curious: why not switch to immutables?

jared2501 replied:
outside the scope of this change :)

jared2501 replied:
Also this is more groovy!

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/ProductId.groovy:
```diff
@@ -0,0 +1,31 @@
+import groovy.transform.EqualsAndHashCode
+
+@EqualsAndHashCode
+class ProductId implements Serializable {
```
why not use immutables?

jared2501 replied:
I didn't need JSON ser/de on it

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/RecommendedProductDependencies.java:
```diff
@@ -0,0 +1,25 @@
+@JsonDeserialize(as = ImmutableRecommendedProductDependencies.class)
+public interface RecommendedProductDependencies {
+
+    @JsonProperty("recommended-product-dependencies")
```
curious: why kebob?

jared2501 replied:
things that are in config are generally kebab case, i also coppied "manifest-extensions" - no pref here though, can change

jmcampanini replied:
figured it's worth lining up with the manifest conventions. dunno what that is atm though :)

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/service/JavaServiceDistributionPlugin.groovy:
```diff
@@ -77,15 +77,18 @@ class JavaServiceDistributionPlugin implements Plugin<Project> {
             checkScript.configure(distributionExtension.serviceName, distributionExtension.checkArgs)
         }
 
+        distributionExtension.productDependenciesConfig = project.configurations.getByName("runtime")
```
if done when applying plugin, then the configuration must exist, meaning the java plugin must have already added it. if there is a way to delay this or evaluate it lazily, that would be ideal. maybe move to right above line 84, in the afterEvaluate?

jared2501 replied:
The JavaServiceDistributionPlugin applies the java plugin

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -76,23 +193,30 @@ class CreateManifestTask extends DefaultTask {
+    }
+
+    static String identifierToCoord(ModuleVersionIdentifier identifier) {
+        return "$identifier.group:$identifier.name:$identifier.version"
```
nit: use `${}` notation, to make the variable interpolation easier to read

jared2501 replied:
done

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+        Map<String, String> mavenCoordsByProductIds = [:]
+
+        productDependenciesConfig.resolvedConfiguration.resolvedArtifacts.each { artifact ->
+            def coord = identifierToCoord(artifact.moduleVersion.id)
```
nit: explicit type for `coord`

jared2501 replied:
done (note IJ does type inference on defs)

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+            def recommendedDeps = jsonMapper.readValue(pdeps, RecommendedProductDependencies.class)
+
+            if (!allRecommendedDepsByCoord.containsKey(coord)) {
+                allRecommendedDepsByCoord.put(coord, new HashSet<RecommendedProductDependency>())
```
nit: multimap may be useful here to minimize boilerplate

jared2501 replied:
Then I'd need to pull in guava as a dependency :/ have had issue in the past when plugins have done this... your call

jmcampanini replied:
yup. i understand guava being a source of pain on the buildscript classpath

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+                def productId = "${recommendedDep.productGroup()}:${recommendedDep.productName()}".toString()
+                if (mavenCoordsByProductIds.containsKey(productId)) {
+                    def othercoord = mavenCoordsByProductIds.get(productId)
+                    throw new GradleException("Duplicate product dependency recommendations found for " +
```
1. no test for this?
2. this would happen if 2 different JARs recommended the same product, correct? if so, how does that work with 2 products having a shared dependency (diamond)

jared2501 replied:
@jmcampanini -

1. Yeah no test, can add one if you want, but just got bored of writing tests...
2. Only API jars should recommend product dependencies. Gradle will have to conflict resolve the API jars to a single jar anyway, so this should never be a problem

jmcampanini replied:
ah yes, it's after the resolution. 👍

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+        Map<String, RecommendedProductDependency> recommendedDepsByProductId = [:]
+        allRecommendedProductDeps.each { recommendedDep ->
+            def productId = "${recommendedDep.productGroup()}:${recommendedDep.productName()}".toString()
+            recommendedDepsByProductId.put(productId, recommendedDep)
```
can this be done next to line 132?

jared2501 replied:
yup, done

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+
+                } catch (IllegalArgumentException e) {
+                    def mavenCoordSource = mavenCoordsByProductIds.get(productId)
+                    throw new GradleException("IllegalArgumentException encountered when generating " +
```
why would this happen?

jared2501 replied:
if the RecommendedProductDependency was invalid (i.e. didn't have min/max/recommended all set)

jmcampanini commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/tasks/CreateManifestTask.groovy:
```diff
@@ -66,6 +85,104 @@ class CreateManifestTask extends DefaultTask {
+        }
+
+        def unseenProductIds = new HashSet<>(recommendedDepsByProductId.keySet())
+        seenRecommendedProductIds.each { unseenProductIds.remove(it) }
```
i think you can do set subtraction in groovy: https://stackoverflow.com/questions/5713665/opposite-of-intersect-in-groovy-collections

jared2501 replied:
hmm yeah i wasn't quite sure if the returned set was mutable, and the line below needs to call toString before set subtraction, your call here

jmcampanini commented on gradle-sls-packaging/src/test/groovy/com/palantir/gradle/dist/service/ServiceDistributionPluginTests.groovy:
```diff
@@ -500,7 +500,7 @@ class ServiceDistributionPluginTests extends GradleTestSpec {
         classpathJar.exists()
         readFromZip(classpathJar, "META-INF/MANIFEST.MF")
-                .contains('Class-Path: guava-19.0.jar produces-manifest-') // etc
+                .contains('Class-Path: guava-19.0.jar root-project-manifest-') // etc
```
is the reason for this change the `settingsFile` addition to `GradleTestKit` ?

jared2501 replied:
the previous version of nebula-test this repo was on didn't generate maven repo when using DependencyGraph with the version of gradle it was on. Therefore, I bumped nebula-test to latest, but the latest version of nebula-test used the crazy temporary folder annotation this projected used before. I switched to junit's TemporaryFolder, but it generates the folder as a UUID, so i had to set rootProject.name in settings.gradle so set it to somethign sane (root-project) and hence this chage...
phew

jmcampanini reviewed (commented):
* you can mix both java + groovy code in same code base. we could evolve this to using java for non-tests slowly
* would like to see some of the developer ergonomics improved (fully qualified names in consumer's build scripts) but not functionality related

Commit:
Address comments

jmcampanini reviewed (approved):
lgtm. only nit is `ignoredProductDependency` is unused

jared2501 commented:
@jmcampanini when I have part 2 of this change (the task that jams the recommended product dependencies in the jar manifest) I can write an end to end test that uses the extension if you fancy? :D

jmcampanini commented:
👍  lgtm. would def like an ete test that uses both parts. and i am fancy ;)

jared2501 commented:
Awesome! @jmcampanini - do you think you can approve + merge this? Will get a part 2 up shortly.

iamdanfox commented on gradle-sls-packaging/src/main/groovy/com/palantir/gradle/dist/BaseDistributionExtension.groovy:
```diff
@@ -83,6 +94,14 @@ class BaseDistributionExtension {
+        this.productDependenciesConfig = productDependenciesConfig
+    }
+
+    void ignoredProductDependency(String productGroup, String productName) {
```
@jared2501 could you mention this in your docs PR? I was pretty surprised a 2.4 -> 2.7 upgrade broke my build due to a pdep from MP!
